releasenotes/notes/expose-qk-complex64-3b0a134669b2def4.yaml

```diff
@@ -0,0 +1,41 @@
+---
+upgrade_c:
+  - |
+    The way complex numbers are exposed in Qiskit's C API has been updgraded.
+    Previously, ``QkComplex64`` was a compiler-dependent typedef that allowed to pass
+    native complex types by pointer to Qiskit's API (for example as ``double complex*``).
+    While this was convenient, this approach implictly relied on memory-layout assumptions
+    that are not strictly guaranteed.
+
+    Qiskit v2.1 now exposes ``QkComplex64 { double re; double im; }`` as struct, to ensure the
+    memory layout is always compatible and for broader compiler support. For convenience,
+    compiler-dependent converters ``qk_complex64_from_native`` and ``qk_complex64_to_native``
+    are provided, which allows translating from the struct to a native complex number.
+    Note that these only work on platforms supporting ``double complex`` or for MSVC compilers
+    using ``_Dcomplex``.
+
+    For example::
+
+      #include <qiskit.h>
+      #include <math.h>
+      #include <stdio.h>
+      #include <complex.h>
+
+      void main() {
+          // using converter
+          double complex native = 5.0 + I * 3; // uses C11 standard, does not work on MSVC
+          QkComplex64 coeff = qk_complex64_from_native(&native);  // convert from native
+
+          // platform-independent constructions:
+          // QkComplex64 coeff = {5.0, 3.0};
+
+          uint32_t num_qubits = 100;
+          QkObs *obs = qk_obs_zero(num_qubits);
+          QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+          uint32_t indices[3] = {0, 1, 2};
+          QkObsTerm term = {coeff, 3, bit_terms, indices, num_qubits};
+          qk_obs_add_term(obs, &term);
+          printf("num_qubits: %u\n", qk_obs_num_qubits(obs));
+          printf("num_terms: %lu\n", qk_obs_num_terms(obs));
```

Comment:
If you care about such things:
1. you don't `free` the observable (though since we just run off the end of `main`, it hardly matters)
2. neither `%u` nor `%lu` are precisely the correct format specifiers for `uint32_t` or `size_t`.  I think `%lu` should always work for `uint32_t` after promotion, but `size_t` is technically _supposed_ to be `%zu`.  I think the _proper proper_ way to do format strings for `uint32_t` in a portable way is to do
   ```c
   #include <inttypes.h>
   #include <stdint.h>
   int main(void) {
       uint32_t num_qubits = 0;
       printf("num_qubits: " PRIu32 "\n", num_qubits);
       return 0;
   }
   ```
   but it's kind of insane.  I don't really care, just saying.

Reply:
1. I do care!
2. I also do care, but not enough to put `PRIu32` 😂 though this is very interesting to know, thanks!